CA1305 specify IFormatProvider

basoundr opened the pull request:
Implements CA1035 Specify IFormatProvider rule

Cause: A method or constructor calls one or more members that have overloads that accept a System.IFormatProvider parameter, and the method or constructor does not call the overload that takes the IFormatProvider parameter. There are some exceptions to this rule where we dont fire diagnostics for `Activator` type members and `ResourceManager` type members.

Tagging @dotnet/roslyn-analysis for review

basoundr commented:
@mavasani @srivatsn ping?

srivatsn commented on src/System.Runtime.Analyzers/Core/SpecifyIFormatProvider.cs:
```diff
@@ -27,42 +34,203 @@ public abstract class SpecifyIFormatProviderAnalyzer : DiagnosticAnalyzer
+                var computerInfoType = csaContext.Compilation.GetTypeByMetadataName("Microsoft.VisualBasic.Devices.ComputerInfo");
+                var installedUICulturePropertyOfComputerInfoType = computerInfoType?.GetMembers("InstalledUICulture").OfType<IPropertySymbol>().SingleOrDefault();
+                #endregion
+
```
Is it possible to skip doing any analysis if any of these types are not present? Or are most of the types only needed for corner cases?

srivatsn replied:
If IFormatProvider is not there then there's no point to running this rule?

mavasani replied:
+1
I think Portable profile doesn't define these types/members, so we should check for null well known types upfront.

basoundr replied:
will do

srivatsn replied:
Where? I don't see any checks for before registering the operationaction

srivatsn replied:
Never mind, I was looking at a stale version.

srivatsn commented on src/System.Runtime.Analyzers/Core/SpecifyIFormatProvider.cs:
```diff
@@ -27,42 +34,203 @@ public abstract class SpecifyIFormatProviderAnalyzer : DiagnosticAnalyzer
+                         targetMethod.Equals(stringFormatMemberWithStringObjectObjectAndObjectParameter) ||
+                         targetMethod.Equals(stringFormatMemberWithStringAndParamsObjectParameter)))
+                    {
+                        // Sample message for IFormatProviderAlternateStringRule: Because the behavior of string.Format(string, object) could vary based on the current user's locale settings,
```
Wow these messages are wordy. Does it look like a block of unintelligible text in the error list? Is there an opportunity to simplify these and shove stuff into the description instead?

basoundr replied:
These messages are wordy but they are actionable providing information on what a replacement could be based on the user scenario.

srivatsn commented on src/System.Runtime.Analyzers/Core/SpecifyIFormatProvider.cs:
```diff
@@ -27,42 +34,203 @@ public abstract class SpecifyIFormatProviderAnalyzer : DiagnosticAnalyzer
+                        oaContext.ReportDiagnostic(
+                            invocationExpression.Syntax.CreateDiagnostic(
+                                IFormatProviderAlternateStringRule,
+                                targetMethod.ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat),
```
Isn't the CSharpErrorMessageFormat the default? Do you have explicitly specify it?

basoundr replied:
It works well if the symbols come from CSharp Compilation but they are from the VBCompilation then the display changes to VB format. To prevent that, we need to use this

srivatsn replied:
But for a VB project, we should format the message according to the VB format right? Why do we want to force it to be the C# style?

basoundr replied:
The default display of the VB symbol is not user-friendly. Here is the display of the `CurrentCulture` Property display from the Type `CultureInfo` for VB and C#

VB: `Public Shared Overloads Property CurrentCulture As System.Globalization.CultureInfo`
C#: `System.Globalization.CultureInfo.CurrentCulture`

Now imagine this, as part of the message in the commented code section.

Also `CSharpErrorMessageFormat`, though has the name `CSharp` in it, has nothing specific to C#. It is just a `SymbolDisplayFormat` that is used for displaying C# symbols. We can use it to display any symbol in that desirable format.

srivatsn replied:
Ah ok.. In general, I'd say it's probably best to have a more descriptive name for the display format. I remember either @lgolding or @mavasani had run into this before and had created a SymbolDisplayFormat for VB that was more appropriate.

srivatsn replied:
Actually thinking about this a bit more - do we need the fully qualified names here? We went through the compiler and reduced all names in error to simple names to shorten the messages because most often the qualifying it is unnecessary - for eg CultureInfo is not ambiguous at all given the context of the rule. I think you could simply just use symbol.Name

basoundr replied:
`CultureInfo` itself is not ambiguous but there are places where we mention `InstalledUICulture`, a property of `CultureInfo`. This could either be `System.Globalization.CultureInfo.InstalledUICulture` or `Microsoft.VisualBasic.Devices.ComputerInfo.InstalledUICulture`. You can checkout the tests where you will find this.

srivatsn commented on src/System.Runtime.Analyzers/Core/SpecifyIFormatProvider.cs:
```diff
@@ -27,42 +34,203 @@ public abstract class SpecifyIFormatProviderAnalyzer : DiagnosticAnalyzer
+                        // If there are two matching overloads, one with CultureInfo as the first parameter and one with CultureInfo as the last parameter,
+                        // report the diagnostic on the overload with CultureInfo as the last parameter, to match the behavior of FxCop.
+                        var correctOverload = correctOverloads
+                                              .Where(overload => overload.Parameters.Last().Type.Equals(IFormatProviderType))
```
If overload here was a parameterless method would this crash?

basoundr replied:
One of the checks in `GetMethodOverloadsWithDesiredParameterAtLeadingOrTrailing` is to make sure the overload exactly +1 no of parameters than the target method. So this can never be empty.

srivatsn commented on src/System.Runtime.Analyzers/Core/SpecifyIFormatProvider.cs:
```diff
@@ -27,42 +34,203 @@ public abstract class SpecifyIFormatProviderAnalyzer : DiagnosticAnalyzer
+        private IEnumerable<int> GetIndexesOfParameterType(IMethodSymbol targetMethod, INamedTypeSymbol formatProviderType)
+        {
+            return targetMethod.Parameters
+                .Select((Parameter, Index) => new { Parameter, Index })
```
Don't capitalize local names. Curious how dows this work? Isn't targetMethod.Parameters just an array of parametersymbols? Where does index come from?

basoundr replied:
This is a overload of `select`, `public static IEnumerable<TResult> Select<TSource, TResult>(this IEnumerable<TSource> source, Func<TSource, int, TResult> selector);` which provides the index of the parameter inside the `IEnumerable`. 

I capitalized `Parameter` and `Index` because they eventually become Properties of anonymous objects. If you feel the variable names should `camelCased` I can do it. I am fine with either.

srivatsn commented on src/Test/Utilities/DiagnosticAnalyzerTestBase.cs:
```diff
@@ -22,8 +23,9 @@ public abstract class DiagnosticAnalyzerTestBase
         private static readonly MetadataReference s_systemXmlReference = MetadataReference.CreateFromAssemblyInternal(typeof(System.Xml.XmlDocument).Assembly);
         private static readonly MetadataReference s_systemXmlDataReference = MetadataReference.CreateFromAssemblyInternal(typeof(System.Data.Rule).Assembly);
-        private static readonly MetadataReference s_CSharpSymbolsReference = MetadataReference.CreateFromAssemblyInternal(typeof(CSharpCompilation).Assembly);
+        private static readonly MetadataReference s_cSharpSymbolsReference = MetadataReference.CreateFromAssemblyInternal(typeof(CSharpCompilation).Assembly);
```
odd naming style - should it be s_csharpSymbolsReference?

basoundr replied:
Sure

srivatsn commented on src/Analyzer.Utilities/Extensions/IEnumerableOfIMethodSymbolExtensions.cs:
```diff
@@ -0,0 +1,54 @@
+{
+    public static class IEnumerableOfIMethodSymbolExtensions
+    {
+        public static IEnumerable<IMethodSymbol> GetMethodOverloadsWithDesiredParameterAtLeadingOrTrailing(
```
Can you add a doc comment summarizing what this method does.

basoundr replied:
Added comments

basoundr commented:
@dotnet-bot retest default please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn-analyzers_prtest/797/
// Retest reason: Unknown failure

basoundr commented:
@dotnet-bot retest Windows Release please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn-analyzers/job/master/job/windows_release_prtest/87/
// Retest reason: Nuget failure

basoundr commented:
@dotnet/roslyn-analysis good to checkin?

basoundr commented:
@mavasani Now the overload is calculated using Symbol information instead of string comparison :). For the time being, the helper will be in the analyzer itself. I will need to use it in the other PR as well. I will move this to a helper when I modify that PR #915 
@srivatsn I have shortened the message like you suggested and moved that information over to the description.

Hopefully I have addressed all the issues :)

basoundr commented:
@dotnet-bot retest default please
// Previous failure: http://dotnet-ci.cloudapp.net/job/dotnet_roslyn-analyzers_prtest/802/
// Retest reason: java.io.IOException: Connection aborted: org.jenkinsci.remoting.nio.NioChannelHub$MonoNioTransport@233bd08

basoundr commented:
@jaredpar The default queue has failed because of some jenkin's IO issue. Asking to retest isn't working. Is this related to the result caching mechanism that you have in Jenkins? Is there any work to force build it and skip the cache?

jaredpar commented:
@basoundr there is no caching on this repo.  Drop by and I'll show you how to force a retest on this.

basoundr commented:
@srivatsn @mavasani I merged in this PR instead of merging #915. I will address comments on this PR separately
